### PR TITLE
Suggest yarn when installed with yarn

### LINKED ISSUE #118: npm is suggested when installed via yarn (global install only)
I have xo installed via yarn not npm which means if I "update" with npm I'll end up with two copies of xo in my `$PATH`.

```
➜  SickRage git:(develop) ✗ realpath $(which xo)      
/Users/xo/.config/yarn/global/node_modules/xo/cli.js
```

```
   ╭───────────────────────────────────────╮
   │                                       │
   │   Update available 0.17.1 → 0.18.2    │
   │       Run npm i -g xo to update       │
   │                                       │
   ╰───────────────────────────────────────╯
```

### LINKED ISSUE #146: Detect if the CLI has been installed globally or as a dev dependency
When notified to update a CLI, the suggested command the user should run to update the CLI ignores the way the CLI has been installed, e.g. gloablly or as a dev dependency. The suggested command should be adjusted accordingly.

### PR DESCRIPTION
**Pull request by LitoMore**
Fixes #118 
Resolve #146

### CODE REVIEW

**Comment by sindresorhus** [comment]
This is not what was suggested in https://github.com/yeoman/update-notifier/issues/118#issuecomment-315040151 This is not going to be accepted if it requires any `fs` calls. That's not worth the overhead.

File: index.js [hunk 1]
```diff
@@ -107,15 +111,42 @@ class UpdateNotifier {
...
+				cliName = Object.keys(bin)[0];
+				break;
+			case 'undefined':
+			default:
+				return false;
+		}
+
+		try {
+			const realpath = fs.realpathSync(which().sync(cliName, {nothrow: true}));
+			if (realpath.match(/yarn\/global/)) {
```
Comment (nexdrew) [inline index.js]: I could be wrong, but this literal RegExp doesn't look to be Windows-friendly (hardcoded forward slash).

Perhaps we could simplify this to:

```js
return realpath.includes(path.join('yarn', 'global'));
```

**Comment by LitoMore** [comment]
@sindresorhus Thank you for your advice.

`process.cwd()` is a better way to get the path.

File: index.js (hunk 1 above)
Comment (LitoMore) [reply index.js]: Thanks, I will try it in Windows.

File: index.js [hunk 2]
```diff
@@ -107,15 +111,42 @@ class UpdateNotifier {
...
+			return false;
+		}
+	}
 	notify(opts) {
 		if (!process.stdout.isTTY || isNpm() || !this.update) {
 			return this;
 		}
 
 		opts = Object.assign({isGlobal: isInstalledGlobally()}, opts);
+		const installCommand = this.isYarn() ? 'yarn global upgrade ' + this.packageName : 'npm i ' + (opts.isGlobal ? '-g ' : '') + this.packageName;
```
[the remark this answers is not in the document]
Comment (LitoMore) [reply index.js]: We don't have to.

**Comment by LitoMore** [comment]
If the method `isYarnGlobal()` approved, should we create a new package `is-yarn-global` then require it?

File: index.js [hunk 3]
```diff
@@ -107,15 +107,24 @@ class UpdateNotifier {
 			};
 		});
 	}
+	isYarnGlobal() {
+		const isWindows = process.platform === 'win32' || process.env.OSTYPE === 'cygwin' || process.env.OSTYPE === 'msys';
```
Comment (sindresorhus) [inline index.js]: Should only be `const isWindows = process.platform === 'win32';`

File: index.js [hunk 4]
```diff
@@ -107,15 +107,24 @@ class UpdateNotifier {
...
+			return true;
+		}
+		return false;
+	}
 	notify(opts) {
 		if (!process.stdout.isTTY || isNpm() || !this.update) {
 			return this;
 		}
 
 		opts = Object.assign({isGlobal: isInstalledGlobally()}, opts);
```
Comment (sindresorhus) [inline index.js]: `isInstalledGlobally` already supports Yarn, but it doesn't return which one it is. For that, you could use https://github.com/sindresorhus/global-dirs#packages

Comment (LitoMore) [reply index.js]: I created a package [is-yarn-global](https://github.com/LitoMore/is-yarn-global), it without any `fs` calls and dependencies. What about this?

File: index.js (hunk 3 above)
Comment (LitoMore) [reply index.js]: 👌

File: index.js [hunk 5]
```diff
@@ -107,15 +107,24 @@ class UpdateNotifier {
...
+		const installCommand = this.isYarnGlobal() ? 'yarn global upgrade ' + this.packageName : 'npm i ' + (opts.isGlobal ? '-g ' : '') + this.packageName;
```
Comment (LitoMore) [inline index.js]: `yarn global upgrade <package>` will not upgrade package to latest version. Should be `yarn global add <package>` or `yarn global upgrade <package>@latest`.

**Comment by LitoMore** [comment]
@sindresorhus I could be wrong, I think [is-yarn-global](https://github.com/LitoMore/is-yarn-global) is better than [global-dirs](https://github.com/sindresorhus/global-dirs#packages), `is-yarn-global` it without any `fs` calls and new dependencies. Use `is-yarn-global` or `global-dirs` it's up to you.

**Comment by sindresorhus** [comment]
> I could be wrong, I think is-yarn-global is better than global-dirs, is-yarn-global it without any fs calls and new dependencies. Use is-yarn-global or global-dirs it's up to you.

Good point, yes.

**Comment by LitoMore** [comment]
@sindresorhus Changes already pushed, please review :)

**Comment by LitoMore** [comment]
ping @sindresorhus

**Comment by LitoMore** [comment]
@sindresorhus `is-yarn-global` has been landed on this branch, please review :)

**Comment by LitoMore** [comment]
ping @SBoudrias

**Comment by LitoMore** [comment]
@sindresorhus Code updated, please review

**Comment by LitoMore** [comment]
ping @SBoudrias @sindresorhus I updated my code, but the review status does not change.

Any progress with this pull request?

**Comment by LitoMore** [comment]
@sindresorhus Hey Sindre! Please review! Thanks!

**Comment by joshmanders** [comment]
Any update on this?

**Comment by sindresorhus** [comment]
This got lost in my review queue.

@LitoMore Can you fix the merge conflicts?

**Comment by LitoMore** [comment]
@sindresorhus Code updated.

File: index.js [hunk 6]
```diff
@@ -124,11 +125,13 @@ class UpdateNotifier {
 
 		options = {
 			isGlobal: isInstalledGlobally(),
+			isYarnGlobal: isYarnGlobal()(),
 			...options
 		};
+		const installCommand = options.isYarnGlobal ? `yarn global add ${this.packageName}` : `npm i ${options.isGlobal ? '-g ' : ''}${this.packageName}`;
```
Comment (sindresorhus) [inline index.js]: Shouldn't we also show local Yarn command?

Comment (LitoMore) [reply index.js]: > Shouldn't we also show local Yarn command?
@sindresorhus How do we know if it is using Yarn or NPM locally?

Comment (sindresorhus) [reply index.js]: https://github.com/sindresorhus/has-yarn

**Comment by LitoMore** [comment]
@sindresorhus Code updated.

File: index.js [hunk 7]
```diff
@@ -126,11 +128,17 @@ class UpdateNotifier {
 
 		options = {
 			isGlobal: isInstalledGlobally(),
+			isYarnGlobal: isYarnGlobal()(),
 			...options
 		};
+		const installCommand = options.isYarnGlobal ?
```
Comment (SBoudrias) [inline index.js]: Let's change this nested ternary to `if`/`else`:

```js
let installCommand;
if (options.isYarnGlobal) {
   installCommand = `yarn global add ${this.packageName}`;
} else if (...
```

**Comment by bluelovers** [comment]
current path is yarn project not mean this cli install by current local

should not `Suggest yarn add`

image

**Pull request merged**